### PR TITLE
[fx] refactor memory utils and extend shard utils.

### PR DESCRIPTION
**Pull request by super-dainiu**
### Updates
I believe the memory utils and shard utils should be separated for sake of clarity. I also left a TODO in the `shard_utils.py` for future development.

I have been refactoring existing code for days, but I believe I should set apart these changes because too many changes have been made. I hope this will not confuse the reviewers.

### Tests
image

### CODE REVIEW

File: colossalai/fx/passes/algorithms/ckpt_solver_chen.py [hunk 1]
```diff
@@ -1,7 +1,9 @@
+import math
 from typing import List, Set, Tuple
+
 import torch
 from torch.fx import GraphModule, Node
-import math
+
```
Comment (super-dainiu) [inline colossalai/fx/passes/algorithms/ckpt_solver_chen.py:6]: These were done by `pre-commit` hook

File: colossalai/fx/passes/algorithms/ckpt_solver_rotor.py [hunk 2]
```diff
@@ -1,15 +1,17 @@
...
-from .linearize import linearize
-from .operation import ForwardCheck, ForwardEnable, ForwardNograd, Backward, Loss, Chain, Sequence, Function
+
 from colossalai.fx.codegen.activation_checkpoint_codegen import _find_nested_ckpt_regions
+from colossalai.fx.graph_module import ColoGraphModule
+from colossalai.fx.profiler import activation_size, calculate_fwd_out, calculate_fwd_tmp, parameter_size
 from colossalai.logging import get_dist_logger
 
+from .linearize import linearize
+from .operation import Backward, Chain, ForwardCheck, ForwardEnable, ForwardNograd, Function, Loss, Sequence
```
Comment (super-dainiu) [inline colossalai/fx/passes/algorithms/ckpt_solver_rotor.py:13]: fix these import

File: colossalai/fx/passes/algorithms/ckpt_solver_rotor.py [hunk 3]
```diff
@@ -18,7 +20,7 @@
 # https://gitlab.inria.fr/hiepacs/rotor
 # paper link: https://hal.inria.fr/hal-02352969
 def _compute_table(chain: Chain, mmax) -> Tuple:
-    """Returns the optimal table: a tuple containing: 
+    """Returns the optimal table: a tuple containing:
     Opt[m][lmin][lmax] with lmin = 0...chain.length
```
Comment (super-dainiu) [inline colossalai/fx/passes/algorithms/ckpt_solver_rotor.py:24]: Don't know why but the new `pre-commit` hook is so rigid

File: colossalai/fx/passes/concrete_info_prop.py [hunk 4]
```diff
@@ -3,11 +3,12 @@
 
 import torch
 import torch.fx
```
Comment (super-dainiu) [inline colossalai/fx/passes/concrete_info_prop.py:5]: Also nothing changed in this file

File: colossalai/fx/passes/meta_info_prop.py [hunk 5]
```diff
@@ -3,12 +3,21 @@
...
+from colossalai.fx.profiler import (
+    GraphInfo,
+    activation_size,
+    calculate_fwd_in,
+    calculate_fwd_out,
+    calculate_fwd_tmp,
+    profile_function,
+    profile_method,
+    profile_module,
+)
```
Comment (super-dainiu) [inline colossalai/fx/passes/meta_info_prop.py:19]: Only changes import

File: colossalai/fx/profiler/__init__.py [hunk 6]
```diff
@@ -1,12 +1,12 @@
 from .._compatibility import is_compatible_with_meta
 
 if is_compatible_with_meta():
-    from .memory import calculate_fwd_in, calculate_fwd_out, calculate_fwd_tmp
     from .opcount import flop_mapping
     from .profiler import profile_function, profile_method, profile_module
+    from .shard_utils import calculate_fwd_in, calculate_fwd_out, calculate_fwd_tmp
```
Comment (super-dainiu) [inline colossalai/fx/profiler/__init__.py:6]: only changes import

File: colossalai/fx/profiler/dataflow.py [hunk 7]
```diff
@@ -6,7 +6,7 @@
 from torch.fx import Graph, Node
 
 from .._compatibility import compatibility
-from .memory import activation_size, is_inplace
+from .memory_utils import activation_size, is_inplace
```
Comment (super-dainiu) [inline colossalai/fx/profiler/dataflow.py:9]: only changes import

File: colossalai/fx/profiler/memory_utils.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    if n.op == "call_function":
+        inplace = n.kwargs.get("inplace", False)
+        if is_compatible_with_meta():
+            from .constants import ALIAS_ATEN
+            if n.target in ALIAS_ATEN:
+                inplace = True
+    elif n.op == "call_module":
+        inplace = getattr(n.graph.owning_module.get_submodule(n.target), "inplace", False)
+
+    return inplace
```
Comment (super-dainiu) [inline colossalai/fx/profiler/memory_utils.py:71]: I split `memory.py` into `memory_utils.py` and `shard_utils.py`

File: colossalai/fx/profiler/memory_utils.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+    Returns:
+        int: The activation size
+    """
+    act_size = 0
+    if isinstance(out, torch.Tensor):
+        if out.is_quantized:
+            act_size += out.numel() * torch._empty_affine_quantized([], dtype=out.dtype).element_size()
+        else:
+            act_size += out.numel() * torch.tensor([], dtype=out.dtype).element_size()
```
Comment (super-dainiu) [inline colossalai/fx/profiler/memory_utils.py:26]: here I consider the quantized version

File: colossalai/fx/profiler/shard_utils.py [hunk 10]
```diff
@@ -1,70 +1,32 @@
-from typing import Dict, List, Tuple, Union
-
 import torch
-from torch.fx import GraphModule, Node
+from torch.fx import Node
 
 from .._compatibility import compatibility, is_compatible_with_meta
```
Comment (super-dainiu) [inline colossalai/fx/profiler/shard_utils.py:4]: This file is split from `memory.py`. It's a renaming.

File: colossalai/fx/profiler/shard_utils.py [hunk 11]
```diff
@@ -1,70 +1,32 @@
...
-    """A helper function to calculate `fwd_in`
+    """A helper function to calculate `fwd_in` (with sharding spec)
 
     Args:
         n (Node): a node from the graph
 
     Returns:
         fwd_in (int): the result of `fwd_in`
     """
+    # TODO(super-dainiu): should divide the memory by sharding spec
```
Comment (super-dainiu) [inline colossalai/fx/profiler/shard_utils.py:23]: I leave the todo here. In the future, we will dive deep into sharding spec within this function.

File: tests/test_fx/test_profiler/test_profiler_meta_info_prop.py [hunk 12]
```diff
@@ -3,12 +3,13 @@
 import torch
 import torch.fx
 import torchvision.models as tm
+from gpt_utils import gpt2_medium, gpt2_xl
```
Comment (super-dainiu) [inline tests/test_fx/test_profiler/test_profiler_meta_info_prop.py:6]: only changes import

**Review by Cypher30 — APPROVED** [review]

**Pull request merged**